libtenzir/src/table_slice.cpp

```diff
@@ -1302,4 +1301,241 @@ auto unflatten(const table_slice& slice,
+        | std::views::transform([](auto subrange) {
+            return std::string_view{subrange.data(), subrange.size()};
+          });
+    auto current = &root;
+    for (auto segment : segments) {
+      auto record = std::get_if<unflatten_record>(current);
+      if (record) {
+        // nullptr -> all valid. we only want to make more valid.
+        record->null_bitmap
+          = bitmap_or(record->null_bitmap, array.null_bitmap());
+      } else {
+        // TODO: Get null bitmap of all ones?
+        record = &current->emplace<unflatten_record>(array.length(),
+                                                     array.null_bitmap());
+      }
+      current = &record->fields[segment];
+    }
+    // TODO: data might be record
+    unflatten_recurse(*current, data, sep);
+  }
+}
+
+void unflatten_recurse(unflatten_entry& entry,
+                       std::shared_ptr<arrow::Array> array,
+                       std::string_view sep) {
+  if (auto record = caf::get_if<arrow::StructArray>(&*array)) {
+    unflatten_recurse(entry, *record, sep);
+  } else if (auto list = caf::get_if<arrow::ListArray>(&*array)) {
+    entry = unflatten(*list, sep);
+  } else {
+    entry = std::move(array);
+  }
+}
+
+auto realize(unflatten_record&& record) -> std::shared_ptr<arrow::StructArray>;
+
+auto realize(unflatten_entry&& entry) -> std::shared_ptr<arrow::Array> {
+  TENZIR_WARN("pre realize");
+  return entry.match<std::shared_ptr<arrow::Array>>(
+    [](std::shared_ptr<arrow::Array>& array) {
+      TENZIR_ASSERT(array);
+      TENZIR_WARN("realizing {}", array->type()->ToString());
+      return std::move(array);
+    },
+    [](unflatten_record& record) {
+      TENZIR_WARN("realizing record");
+      return realize(std::move(record));
+    });
+}
+
+auto realize(unflatten_record&& record) -> std::shared_ptr<arrow::StructArray> {
+  TENZIR_WARN("realizing record now");
+  auto names = std::vector<std::string>{};
+  auto arrays = std::vector<std::shared_ptr<arrow::Array>>{};
+  for (auto& [name, entry] : record.fields) {
+    TENZIR_WARN("seeing {}", name);
+    names.emplace_back(name);
+    arrays.push_back(realize(std::move(entry)));
+  }
+  return make_struct_array(record.length, record.null_bitmap, std::move(names),
+                           arrays);
+}
+
+auto unflatten(const arrow::StructArray& array, std::string_view sep)
+  -> std::shared_ptr<arrow::StructArray> {
+  // { foo.bar.baz: 1, foo.bar.qux: 2, foo: {bar: 42} }
+  auto root = unflatten_entry{unflatten_record{array.length(), nullptr}};
+  unflatten_recurse(root, array, sep);
+  auto record = std::get_if<unflatten_record>(&root);
+  TENZIR_ASSERT(record);
+  return realize(std::move(*record));
+}
+
+auto unflatten(std::shared_ptr<arrow::Array> array, std::string_view sep)
+  -> std::shared_ptr<arrow::Array>;
+
+auto unflatten(const arrow::ListArray& array, std::string_view sep)
+  -> std::shared_ptr<arrow::ListArray> {
+  // TODO: array.offset?
```

Comment:
Shouldn't that be handled by `array.values()`, or does that not take care of the offset?

Reply:
Apparently not:

```cpp
/// \brief Return array object containing the list's values
///
/// Note that this buffer does not account for any slice offset or length.
const std::shared_ptr<Array>& values() const { return values_; }
```